rfcs/proposed/task_group_dynamic_dependencies/extended_semantics.md

```diff
@@ -0,0 +1,1453 @@
+on the task `12` and should not wait until `14` to complete.
+
+The improved approach is described in the [Cache-Oblivious Wavefront](https://people.csail.mit.edu/yuantang/Docs/PPoPP15-yuantang.pdf) paper as eager recursion.
+
+In such a recursion each dividing task knows that it's dependent tasks would split as well and notifies them about the sub-tasks required to make only the real
+dependencies between tasks.
+
+Let's consider a wavefront calculations on the square grid ``[0, n)`` on ``x`` and ``y`` dimensions.
+
+Similarly to the classic algorithm, the eager one splits the calculated area into 4 sub-areas by creating 4
+subtasks - ``north_subtask``, ``wast_subtask``, ``east_subtask`` and ``south_subtask``.
+
+The algorithm  also sets the same dependencies as a classic one:
+* ``west_subtask`` and ``east_subtask`` are successors of ``north_subtask``
+* ``south_subtask`` is a successor of ``west_subtask`` and ``east_subtask``.
+
+During the first division, no extra actions needed since the task that splits calculation of the
+whole grid ``[0, n))`` can't have successors.
+
+But during the second division - when the north, west, east and south are executed, each task needs to notify it's
+successors about the subtasks to make dependencies.
+
+Let's look into the second division in details. Initially, we have 4 subtasks created during the first division:
+
+<img src="assets/wavefront_recursive_eager_first_division.png" width=300>
+
+The north subtask is executed first because dependencies prevents west, east and south tasks from executing.
+Similarly to the previous step, it creates the north, east, west and south subtasks:
+
+<img src="assets/wavefront_recursive_eager_second_division_north.png">
+
+The second step is to notify the successors about the subtasks:
+* ``W`` should be notified about west and south subtasks (``NW`` and ``NS``),
+* ``E`` should be notified about east and south subtasks (``NE`` and ``NS``).
+
+Notifying the successor is made by creating ``task_tracker`` objects tracking the corresponding tasks and storing them
+in the container visible for the successors.
+
+North subtask ``NN`` can start execution and division after completing the split of the previous level and notifying
+the successors.
+
+Since ``N`` task is completed, ``W`` and ``E`` can start executing. ``W`` would also split itself into subtasks and
+notify ``S`` about the successors:
+
+<img src="assets/wavefront_recursive_eager_second_division_west.png">
+
+Additionally, ``W`` should add dependencies between it's subtasks ``WN`` and ``WE`` and the subtasks of ``N`` created
+on the previous step (``NW`` and ``NS``):
+
+<img src="assets/wavefront_recursive_eager_second_division_west_new_dependencies.png">
+
+It is important to mention that at the moment of making additional edges, ``NW`` and ``NS`` are at least
+in `submitted` state and even can be completed. 
```

Comment:
Could you please explain why `NW` and `NS` are at least in `submitted` state? It is not obvious to me.